src/protocols/der/decode.c

```diff
@@ -0,0 +1,2852 @@
+
+	size_t len = fr_dbuff_remaining(&our_in);
+
+	if (!fr_type_is_octets(parent->type)) {
+		fr_strerror_printf("Octetstring found in non-octets attribute %s of type %s", parent->name,
+				   fr_type_to_str(parent->type));
+		return -1;
+	}
+
+	/*
+	 *	ISO/IEC 8825-1:2021
+	 *	8.7 Encoding of an octetstring value
+	 *		8.7.1 The encoding of an octetstring value shall be either primitive or constructed at the
+	 *		      option of the sender.
+	 *			NOTE – Where it is necessary to transfer part of an octet string before the entire
+	 *			       octetstring is available, the constructed encoding is used.
+	 *		8.7.2 The primitive encoding contains zero, one or more contents octets equal in value to the
+	 *		      octets in the data value, in the order they appear in the data value, and with the most
+	 *		      significant bit of an octet of the data value aligned with the most significant bit of an
+	 *		      octet of the contents octets.
+	 *		8.7.3 The contents octets for the constructed encoding shall consist of zero, one, or more
+	 *		      encodings.
+	 *			NOTE – Each such encoding includes identifier, length, and contents octets, and may
+	 *			       include end-of-contents octets if it is constructed.
+	 *			8.7.3.1 To encode an octetstring value in this way, it is segmented. Each segment shall
+	 *			       consist of a series of consecutive octets of the value. There shall be no
+	 *			       significance placed on the segment boundaries.
+	 *				NOTE – A segment may be of size zero, i.e. contain no octets.
+	 *
+	 *	10.2 String encoding forms
+	 *		For bitstring, octetstring and restricted character string types, the constructed form of
+	 *		encoding shall not be used. (Contrast with 8.23.6.)
+	 */
+
+	vp = fr_pair_afrom_da(ctx, parent);
+	if (unlikely(vp == NULL)) {
+		fr_strerror_const("Out of memory for octetstring pair");
+		return -1;
+	}
+
+	if (unlikely(fr_pair_value_mem_alloc(vp, &data, len, false) < 0)) {
+		fr_strerror_const("Out of memory for octetstring");
+		return -1;
+	}
+
+	fr_dbuff_out_memcpy(data, &our_in, len);
+
+	fr_pair_append(out, vp);
+
+	return fr_dbuff_set(in, &our_in);
+}
+
+static ssize_t fr_der_decode_null(TALLOC_CTX *ctx, fr_pair_list_t *out, fr_dict_attr_t const *parent, fr_dbuff_t *in,
+				  UNUSED fr_der_decode_ctx_t *decode_ctx)
+{
+	fr_pair_t *vp;
+	fr_dbuff_t our_in = FR_DBUFF(in);
+
+	if (fr_dbuff_remaining(&our_in) != 0) {
+		fr_strerror_const("Null has non-zero length");
+		return -1;
+	}
+
+	/*
+	 *	ISO/IEC 8825-1:2021
+	 *	8.8 Encoding of a null value 8.8.1 The encoding of a null value shall be primitive. 8.8.2 The contents
+	 *	    octets shall not contain any octets. NOTE – The length octet is zero.
+	 */
+
+	vp = fr_pair_afrom_da(ctx, parent);
+	if (unlikely(vp == NULL)) {
+		fr_strerror_const("Out of memory for null pair");
+		return -1;
+	}
+
+	fr_pair_append(out, vp);
+
+	return fr_dbuff_set(in, &our_in);
+}
+
+typedef struct {
+	TALLOC_CTX	     *ctx; /**< Allocation context */
```

Comment:
comments can be aligned vertically, and use `//!<` which is more consistent with the rest of the code.